[WIP] Distributor concurrency

ramonsmits opened the pull request:
Documented the possibility to increase the maximum concurrency before using the distributor.
Updated to sample to increase the maximum concurrency and usage on both the distributor as worker nodes.

weralabaj commented on nservicebus/scalability-and-ha/distributor/index.md:
```diff
@@ -46,6 +46,8 @@ To get a sense of the expected performance take your regular endpoint performanc
 
 If you need to scale out small units of work you might consider splitting your handlers into smaller vertical slices of functionality and deploying them on their own endpoints.
 
+NOTE: Before scaling out via the distributor make sure that you have [increased the maximum concurrency on your current endpoint](/nservicebus/operations/tuning.md#tuning-throughput). The default concurrency level is set to 1 meaning sequential processing of messages. Also make sure that the **MaximumConcurrencyLevel** is increased on the endpoint that is the distributor node to allow it to forwa ...
```
> The default concurrency level is set to 1 meaning sequential processing of messages.

The default concurrency level is set to 1 thread, which results in sequential processing of messages.

or 

The default concurrency level is set to 1 thread, meaning that messages will be processed sequentially.

SimonCropp replied:
remove "you"

SimonCropp replied:
> The default concurrency level is set to 1, which equates to sequential processing of messages.

SimonCropp replied:
> Ensure that the **MaximumConcurrencyLevel** is increased on the distributor node endpoint to allow it to forward messages to workers in parallel.

ramonsmits replied:
@weralabaj Max threads != Max Concurrency, I think that is the reason why in V4 the name was changed.

@SimonCropp Thanks for the grammar check :-)

weralabaj commented on nservicebus/scalability-and-ha/distributor/index.md:
```diff
@@ -210,4 +212,4 @@ Check out [John Breakwell's blog](http://blogs.msdn.com/b/johnbreakwell/archive/
 
-Since the Distributor does not do CPU or memory intensive work, you can often put several Distributor processes on the same clustered server. Be aware that the network IO may end up being the bottleneck for the Distributor, so take into account message sizes and throughput when sizing your infrastructure.
\ No newline at end of file
+Since the Distributor does not do CPU or memory intensive work, you can often put several Distributor processes on the same clustered server. Be aware that the network IO may end up being the bottleneck for the Distributor, so take into account message sizes and throughput when sizing your infrastructure.
```
> when sizing your infrastructure

I know you didn't write it now, but wouldn't it be better to say "when configuring" or "when fine-tuning" your infrastructure?

SimonCropp replied:
remove "you"

SimonCropp replied:
> Since the Distributor performs **no** CPU or memory intensive work

weralabaj commented on samples/scaleout/distributor/sample.md:
```diff
@@ -67,12 +67,21 @@ The works are simply host for running the handlers defined in `Worker.Handlers`
 
+### Parallel processing of messages
+
+By default NServiceBus (pre V6) processes message sequentially. The following configuration increases the maximum concurrency level to the number of cores of the machine.
```
> pre V6

To be consistent with other docos, say "In Version 5 and below NServiceBus by default processes messages sequentially. (...)"

weralabaj replied:
Is the concurrency level == number of cores the recommended approach or a good starting point? If so I think we should say that explicitly.

weralabaj commented on samples/scaleout/distributor/sample.md:
```diff
@@ -67,12 +67,21 @@ The works are simply host for running the handlers defined in `Worker.Handlers`
+
+snippet: concurrency
+
+NOTE: Increasing the maximum concurrency level on the distributor allows the distributor to forward messages to workers concurrently. Increasing the concurrency on the workers might not lead to increased performance if the executed code is multi threaded.
```
Any limits or downsides of increasing it too much on distributor node?

SimonCropp replied:
this is a sample. but the distributor is being documented here. it would be better to put this content in the distributor article  and link from here. The reason is that if we have multiple distributor samples we would prefer not to duplicate this content

weralabaj commented:
@ramonsmits I had a few comments, if that's not a WIP then after addressing them I'm ready to merge :D

SimonCropp commented on samples/scaleout/distributor/Version_4.3/Shared/ConfigureTransportConfig.cs:
```diff
@@ -0,0 +1,16 @@
+using NServiceBus.Config;
+using NServiceBus.Config.ConfigurationSource;
+
+#region concurrency
```
i think this should be in the doco article not in the sample

ramonsmits commented:
@weralabaj @SimonCropp Yes, the most stuff should probably be put in the distributor doco. To me it makes sense to have the concurrency setting increased in the sample even though this sample is not about throughput but only about the distributor.  Lots of developers will also use this sample to check the 'distributor performance'.

What a good setting is depends on lots of factors. In my experience the sweet spot is usually somewhere between [number of cores] and [number of cores * 3]. It depends on lots of factors, number of endpoints on the machine, duration of handlers, etc.

Maybe instead of this we might need a document about *Increasing and optimizing NServiceBus performance* ?

I agree with @SimonCropp that the *concurrency* snippet should be in the doco

weralabaj commented:
I think that can get huge quickly :) How about a new section in the distribution doco about optimizing its performance and linking to this section from here?

ramonsmits commented:
@weralabaj That is a nice start. I'll make an issue about a performance document as I currently annoyed that if you google on *NServiceBus Performance* that you get the document about our performance counters.

weralabaj commented:
That sounds good to me.

SimonCropp commented:
> To me it makes sense to have the concurrency setting increased in the sample even though this sample is not about throughput but only about the distributor. Lots of developers will also use this sample to check the 'distributor performance'.

@ramonsmits optimizing samples for performance and still maintain the simplicity of them to concisely show the specific of each sample target would be difficult.

Feel free to to link to some other distributor perf article but remove that code from the sample

SimonCropp commented:
@ramonsmits are you still working on this?

indualagarsamy commented:
@ramonsmits - do you need any help with wrapping this PR? Ping me.

ramonsmits commented:
I'll check the comments and either finish it or mention enough here for someone else to continue.

weralabaj commented:
Sorry @ramonsmits just noticed your latest comment, I'm not touching it now :)

weralabaj commented:
ping @ramonsmits

ramonsmits commented:
@SimonCropp @weralabaj @indualagarsamy I've removed the sample changes. Only the section in the distributor doco remains and I've rewritten that.

Please feel free to apply any changes before merging.
